org/mozilla/jss/netscape/security/x509/CertificateChain.java

```diff
@@ -150,6 +153,89 @@ private void readObject(java.io.ObjectInputStream in)
         decode(in);
     }
 
+    public void addCertificate(X509Certificate cert) {
+        certs.add(cert);
+    }
+
+    public void addCertificateChain(CertificateChain certChain) {
+        certs.addAll(certChain.certs);
+    }
+
+    public void addPKCS7(PKCS7 pkcs7) {
+        certs.addAll(Arrays.asList(pkcs7.getCertificates()));
+    }
+
+    /**
+     * Convert a series of PEM certificates or a PKCS #7 data into a certificate chain.
+     * This method will only accept a single chain, so it cannot be used to load CA bundle.
+     */
+    public static CertificateChain fromPEMString(String input) throws Exception {
+
+        CertificateChain certChain = new CertificateChain();
+
+        StringBuilder sb = new StringBuilder();
+        String[] lines = input.split("\\r?\\n");
+
+        for (String line : lines) {
+            line = line.trim();
+
+            if (Cert.HEADER.equals(line)) {
+                sb.setLength(0);
+
+            } else if (Cert.FOOTER.equals(line)) {
+                byte[] bytes = Utils.base64decode(sb.toString());
+                certChain.addCertificate(new X509CertImpl(bytes));
+
+            } else if (PKCS7.HEADER.equals(line)) {
+                sb.setLength(0);
+
+            } else if (PKCS7.FOOTER.equals(line)) {
+                byte[] bytes = Utils.base64decode(sb.toString());
+                PKCS7 pkcs7 = new PKCS7(bytes);
+                certChain.addPKCS7(pkcs7);
+
+            } else {
+                sb.append(line);
+            }
+        }
+
+        // sort and validate the certificate chain
+        certChain.sort();
+
+        return certChain;
+    }
+
+    /**
+     * Convert the certificate chain into a series of PEM certificates.
+     */
+    public String toPEMString() throws Exception {
+
+        StringWriter sw = new StringWriter();
+        PrintWriter pw = new PrintWriter(sw, true);
+
+        for (X509Certificate cert : certs) {
+            pw.println(Cert.HEADER);
+            pw.print(Utils.base64encodeMultiLine(cert.getEncoded()));
+            pw.println(Cert.FOOTER);
+        }
+
+        return sw.toString();
+    }
+
+    /**
+     * Convert the certificate chain into a PKCS #7 object.
+     */
+    public PKCS7 toPKCS7() throws Exception {
+
+        // convert X509Certificate into X509CertImpl
```

Comment:
I think the same comment as above applies here: do we want to sort this chain first? AFAIK, we don't guarantee internally that certs are in sorted order... I don't think it is strictly required either, but we could want it so we could enforce that this is strictly a chain in a CertificateChain class? Up to you.

Reply:
Agreed. I added `sort()` in `toPEMString()` and `toPKCS7()`.